exercises/rollback-with-saga/solution/src/workflows.ts

```diff
@@ -0,0 +1,92 @@
+import { proxyActivities, ApplicationFailure, ActivityFailure, log, sleep } from '@temporalio/workflow';
+import type * as activities from './activities';
+import { compensate, errorMessage } from './compensationUtils';
+import { Distance, PizzaOrder, OrderConfirmation, Compensation } from './shared';
+
+const { sendBill, getDistance, validateCreditCard, refundCustomer, updateInventory, revertInventory } =
+  proxyActivities<typeof activities>({
+    startToCloseTimeout: '5 seconds',
+    retry: {
+      maximumInterval: '10 seconds',
+    },
+  });
+
+export async function pizzaWorkflow(order: PizzaOrder): Promise<OrderConfirmation> {
+  let compensations: Compensation[] = [];
+  let totalPrice = 0;
+  
+  if (order.isDelivery) {
+    let distance: Distance | undefined = undefined;
+
+    try {
+      distance = await getDistance(order.address);
+    } catch (e) {
+      log.error('Unable to get distance', {});
+      throw e;
+    }
+    if (distance.kilometers > 25) {
+      throw ApplicationFailure.create({
+        message: 'Customer lives too far away for delivery',
+        details: [distance.kilometers],
+      });
+    }
+  }
+
+  for (const pizza of order.items) {
+    totalPrice += pizza.price;
+  }
+
+  // Validate the credit card number
+  try {
+    await validateCreditCard(order.customer.creditCardNumber);
+  } catch (err) {
+    if (err instanceof ActivityFailure && err.cause instanceof ApplicationFailure) {
+      log.error(err.cause.message);
+    } else {
+      log.error(`error validating credit card number: ${err}`);
+    }
+  }
+
+  // Add compensation for updating inventory by removing it
```

Comment:
That's too early. Compensation actions should get registered just before calling the "forward" activity. You don't have anything to put back if you never even got to the point of taking it.